src/app/codegen-data-model-provider/EmberDataBuffer.cpp

```diff
@@ -0,0 +1,358 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/codegen-data-model-provider/EmberDataBuffer.h>
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/util/attribute-storage-null-handling.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/TLVTypes.h>
+#include <lib/support/CodeUtils.h>
+#include <protocols/interaction_model/Constants.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+#include <limits>
+
+namespace chip {
+namespace app {
+namespace Ember {
+
+namespace {
+
+/// Maximum length of a string, inclusive
+///
+/// For nullable strings, the maximum length marker is used as a "NULL" marker
+constexpr uint32_t MaxLength(EmberAttributeBuffer::PascalString s, bool nullable)
+{
+    if (s == EmberAttributeBuffer::PascalString::kShort)
+    {
+        return nullable ? std::numeric_limits<uint8_t>::max() - 1 : std::numeric_limits<uint8_t>::max();
+    }
+    // EmberAttributeBuffer::PascalString::kLong:
+    return nullable ? std::numeric_limits<uint16_t>::max() - 1 : std::numeric_limits<uint16_t>::max();
+}
+
+struct UnsignedDecodeInfo
+{
+    unsigned byteCount;
+    uint64_t maxValue;
+};
+
+constexpr UnsignedDecodeInfo GetUnsignedDecodeInfo(EmberAfAttributeType type)
```

Comment:
Code of `{0, 0xFF} and `{0, 0xFFFF}` is super readable. I do not want to pull in more template hoops for this. The point of this is to make the code small too.

I don't think we want to re-use here as there is no `code` really so I do not gain code reuse (flash savings) but I would gain complexity which I do not want.

Reply:
Also existing templates have a `IsNullable` argument. If I use that argument fully I double the table size (with and without nullable) so that is extra code. if I do not use it, I will have to add comments and use a `MaxValue(false)` which I am not convinced will help readability.

Reply:
The point if reuse is that we might need to change the representation here, and the fewer places we have making assumptions about that representation, the better.  This is not about just readability; it's about code maintainability in general.

We should not have multiple sources of truth for how things are represented.  If the existing source of truth does not work here, we should refactor to make it workable.

Reply:
I added code to odd-sized-integers for localization.

The issue I had was that NumericAttributeTraits forces template expansion by adding code that depends on each byte size and is not a single constexpr function, especially on null handling.

The methods I added should be inline constexpr and at least stay in the same file. I somewhat question the `we might need to change the representation` ... we had no proof of this in the last few years of this project, so I would generally apply a YAGNI here becasuse of that.